### PR TITLE
replace line breaks in the app description by spaces - fixes #13315

### LINKED ISSUE #13315: Application description has uneeded line breaks
When the description uses some new-lines as in https://github.com/owncloud/core/blob/543fa0d45b23173ed94f4c865fe38855b1602ee4/apps/provisioning_api/appinfo/info.xml#L5-L13 for a cleaner code the description looks very disturbing with those uneeded line breaks (that is on master):

screen shot 2015-01-13 at 13 34 24

I'm kinda reluctant to change the formatting in `info.xml` as this would make the Git history somewhat cluttersome for edits in the future. Can somebody think of an alternate solution? (maybe searching for `\t` on a new-line?)

### PR DESCRIPTION
**Pull request by MorrisJobke**
cc @DeepDiver1975 @LukasReschke @jancborchardt 

fixes #13315

### CODE REVIEW

**Comment by MorrisJobke** [comment]
The question here is if we should do that for free text elements or just for the description

**Comment by LukasReschke** [comment]
Mhm, actually line breaks are something which are okay for large descriptions. For example the search_lucene apps does it this way: https://github.com/owncloud/search_lucene/blob/eb8d11b80d60c63c530602c67a17ed3f071c65ac/appinfo/info.xml#L13-L44

Way easier to understand than a large text blob.

**Comment by MorrisJobke** [comment]
mmmmh. I also though about that...

**Comment by MorrisJobke** [comment]
So the line break should just be removed if the line also contains non whitespace content?

**Comment by LukasReschke** [comment]
> So the line break should just be removed if the line also contains non whitespace content?

That would be how I imagine it – yes.

**Comment by MorrisJobke** [comment]
So you I really do a preg_replace? ;)

**Comment by jancborchardt** [comment]
I commented on https://github.com/owncloud/core/issues/13315#issuecomment-69752581

**Comment by nickvergessen** [comment]
Maybe add some more magic, like when `\n\n` is used we retain it, but single `\n` are replaced?

**Comment by MorrisJobke** [comment]
> Maybe add some more magic, like when \n\n is used we retain it, but single \n are replaced?

But what is if there is whitespace between the "\n\n". This is a huge regex. It sounds easy but isn't that easy :(

**Comment by MorrisJobke** [comment]
I got an idea how to solve this.

**Comment by MorrisJobke** [comment]
Before:

app-amangement-broken-line-breaks

After:

app-managemnet-proper-line-breaks


I know: Heavy regex. But they are actually more readable then the str_replace stuff and even cached (the cache also tricked me -.-)

**Comment by MorrisJobke** [comment]
I merged the two pictures for better visibility.

**Comment by Xenopathic** [comment]
I agree that multiple line breaks should be preserved, while a single whitespace should be removed. You should be able to use lookbehinds and lookaheads to accomplish this (obviously change the `\n` escaping as necessary:

    preg_replace('(?<=\n)\n(?=\n)', ' ', $subject);

**Comment by Xenopathic** [comment]
You should be able to use lookbehinds and lookaheads:

    preg_replace('(?<!\n)\n(?!\n)', ' ', $subject);

**Comment by Xenopathic** [comment]
This will only run for locally installed apps, since `getAppInfo()` is only run for local apps, while appstore apps bypass it. Perhaps create a new method, `parseAppInfo()`, that is called both by `getAppInfo()` and `getAppstoreApps()`?

File: lib/private/app.php [hunk 1]
```diff
@@ -641,6 +641,28 @@ public static function getAppInfo($appId, $path = false) {
...
 
+		// just modify the description if it is available
+		// otherwise this will create a $data element with an empty 'description'
+		if(isset($data['description'])) {
+			// sometimes the description contains line breaks and they are then also
+			// shown in this way in the app management which isn't wanted as HTML
+			// manages line breaks itself
+
+			// first of all we split on empty lines
+			$paragraphs = preg_split("!\n([[:blank:]]*\n)+!m", $data['description']);
```
Comment (Xenopathic) [inline lib/private/app.php]: You can use `[[:space:]]` to represent both whitespace and newlines, and remove that nasty double qualifier:

    !\n[[:space:]]*\n!m

File: lib/private/app.php [hunk 2]
```diff
@@ -641,6 +641,28 @@ public static function getAppInfo($appId, $path = false) {
...
+
+			$result = [];
+			foreach ($paragraphs as $value) {
+				// replace multiple whitespace (tabs, space, newlines) inside a paragraph
+				// with a single space
+				$result[] = preg_replace('![[:blank:]\\n]+!m', ' ', $value);
```
Comment (Xenopathic) [inline lib/private/app.php]: Same thing here for `[[:space:]]`

**Comment by Xenopathic** [comment]
What happens for apps that use a list format, without blank lines between list elements? For example (from the Video Streamer app):

    It supports:
    1) *.mp4
    2) *.flv
    3) *.wmv(might to ask you to install silverlight)

There's going to be no way around this I fear... perhaps we should think about supporting Markdown descriptions at some point?

**Comment by nickvergessen** [comment]
@Xenopathic see #5322 :)

**Comment by MorrisJobke** [comment]
I now added a new method which is also called for appstore apps. This is just a workaround for #5322

File: lib/private/app.php [hunk 3]
```diff
@@ -1194,4 +1197,34 @@ public static function getStorage($appId) {
 			return false;
 		}
 	}
+
+	/**
+	 * @param $data the app data
+	 * @return the improved app data
+	 */
+	public static function parseAppInfo($data) {
```
Comment (DeepDiver1975) [inline lib/private/app.php]: can we please add unit tests for this ? THX

File: lib/private/app.php [hunk 4]
```diff
@@ -1194,4 +1197,34 @@ public static function getStorage($appId) {
 			return false;
 		}
 	}
+
+	/**
+	 * @param $data the app data
```
Comment (LukasReschke) [inline lib/private/app.php]: string?

Comment (LukasReschke) [inline lib/private/app.php]: or array? - if array add it there and also annotate $data with it for type safeness

Comment (MorrisJobke) [inline lib/private/app.php]: I already thought about that, but it seems that it also can be null. Is this then still a valid annotation?

Comment (LukasReschke) [inline lib/private/app.php]: Sure. `null|array` – but I'd enforce array here and in the functions signature (`function foo(array $data))` and just don't call it for not array values?

**Comment by MorrisJobke** [comment]
Now it contains unit tests :tada: :rocket:

**Comment by Xenopathic** [comment]
Looks good! :+1: I'm still in two minds about the list thing... :ship: for oC 8 or wait for 8.1 (when, hopefully, apps have updated their descriptions?)

**Comment by LukasReschke** [comment]
:+1:

**Pull request merged**